### ISSUE TITLE
Prevent output binary data

### ISSUE THREAD

**Issue by jumale** [Codeception/Codeception#1884]
- **_Known Linux issue:**_ [terminal is broken](http://unix.stackexchange.com/questions/79684/fix-terminal-after-displaying-a-binary-file) after binary data output.
- **_Codeception issue:**_ controllers may return binary response which causes the problem in "verbose" or "debug" mode.
- **_How to reproduce:**_
  - create a controller which returns binary data, for example ZIP archive
  - create a functional test using REST module and send there a GET request to  the controller
  - run the test with -vvv or --debug
- **_Solution:**_ if it's possible, it would be nice to check that response contains binary data, and do not output it [here](https://github.com/Codeception/Codeception/blob/2.0/src/Codeception/Module/REST.php#L396) (maybe show something like string `[Binary Data]` instead).

P.S.: I got this issue in REST module, but probably it can happen in some another modules

**Comment by DavertMik** [comment]
How to do you propose to detect binary data?

**Comment by jumale** [comment]
@DavertMik ok, I've discovered internet and looks like there is no easy solution to define the binary data.
But I found a way how it will be enough for me. All is needed from codeception - allow to define a callback to intercept debug messages. Like that:
```php
<?php
namespace Codeception\Util;

class Debug
{
    /**
     * @var callable
     */
    public static $interceptDebug = null;

    public static function debug($message)
    {
        if (!self::$output) {
            return;
        }
        // modified part
        if (is_callable(static::$interceptDebug)) {
            $message = call_user_func(static::$interceptDebug, $message);
        }
        self::$output->debug($message);
    }

    // ...
}
```

then I can create a helper with the next method:
```php
<?php
public function expectBinaryResponse()
{
    \Codeception\Util\Debug::$interceptDebug = function ($message) {
        if (preg_match('#\[Response\]#', $message)) {
            $message = '[Response] [Binary Data]';
        }
    
        return $message;
    };
}
```

and then in my test:
```php
$I = new Tester($scenario);
$I->expectBinaryResponse();
// ...
```

I know it looks a bit ugly, but I have no other idea how to hack it :)

**Comment by DavertMik** [comment]
Thanks, I will try to think on this too

**Comment by codedokode** [comment]
You could also replace unprintable characters 00-1f with codes like `\x00` so the data are still readable.

**Comment by spikyjt** [comment]
The linked PR did not fix this issue, it only fixed an issue where the request data included binary data.

This issue is about the response having binary data.

I fixed this locally in a similar manner to #2552 with the following at the end of `Codeception\Module\REST::execute()`:

```php
        $response = $this->response;
        if (!ctype_print($response) && false === mb_detect_encoding($response, mb_detect_order(), true)) {
            // if the response has non-printable bytes and it is not a valid unicode string, reformat the
            // display string to signify the presence of a binary response
            $response = '[binary-data length:' . strlen($response) . ' md5:' . md5($response) . ']';
        }
        $this->debugSection("Response", $response);
```

I'll create a PR later if it would be welcomed, but I can't right now as I'm at work!

**Comment by Naktibalda** [comment]
Thanks, please make a PR.

**Comment by spikyjt** [comment]
Done. Sorry for any confusion, I stupidly created a PR on my own fork instead of to the main repo!

**Comment by Naktibalda** [comment]
Fixed by #3985 two years ago.
